### ISSUE TITLE
ASP.NET MiddleWare - Why didn't the designers use an abstract base class or an interface?

### ISSUE THREAD

**Issue by sdg002** [dotnet/AspNetCore.Docs#18219]
[Enter feedback here]
I have just been through the documentation of **ASP.NET Middleware**. I can see functional similarities with **IHttpModule** and **IHttpHandler** of the .NET Framework.

The .NET Core Middleware requires us to implement a custom class with the following :
- A constructor with takes `RequestDelegate` as the first parameter
- A method `InvokeAsync` which takes in a `HttpContext` parameter and returns a Task object 
- Register the middle ware class in the `Startup.cs` by calling `UseMiddleware `method of `IApplicationBuilder`.

I would like to know the thought process that the designers in Microsoft went through to address this challenge.  It is my humble opinion that an interface definition would have better suited the purpose and made the entire process self documenting.

Consider the following hypothetical approach using using an interface  , say IMiddleWare

## Hypothetical interface IMiddleWare
```

interface IMiddleWare
{
    void Init(RequestDelegate next);
    Task InvokeAsync(HttpContext context);
}


```

## Skeletal client side implementation of IMiddleWare

```
class MyMiddleWare : IMiddleWare
{
    RequestDelegate _next;
    ILogger<MyMiddleWare> _logger;

    public MyMidleWare(ILogger<MyMiddleWare> logger)
    {
        //The ctor could accept any other dependencies that have been added to ServiceCollection
        _logger=logger;
    }
    void Init(RequestDelegate next)
    {
        _next=next;
    }

    async Task InvokeAsync(HttpContext context)
    {
      await context.Response.WriteAsync("My Middle ware Test 1");
    }
}


```

## Registering the custom middle ware in the Startup.cs 

```
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers();

           services.AddSingleton<IMiddleWare, MyMiddleWare>();
           services.AddSingleton<IMiddleWare, SomeOtherMiddleWare>();
        }

```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9898482c-fa54-7ce2-1716-d14fbe03990d
* Version Independent ID: 3926c686-0ccc-f9fe-ca52-0134612a1623
* Content: [ASP.NET Core Middleware](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-3.1)
* Content Source: [aspnetcore/fundamentals/middleware/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/fundamentals/middleware/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by Rick-Anderson** [comment]
@sdg002  thanks for the thoughful question. I'm sure they had good reasons. The decision was made many years ago so it's not going to change.

**Comment by davidfowl** [comment]
See https://stackoverflow.com/questions/30442784/is-vnext-completely-convention-based-programming

**Comment by davidfowl** [comment]
Also see https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/extensibility?view=aspnetcore-3.1

**Comment by sdg002** [comment]
Thank you @davidfowl  for pointing me to the right resources. 
I wasn't aware that Microsoft also had an interface based approach to develop custom middleware. 

The experience with the interface based approach was so much more intuitive. Within minutes I was able to create a sample app.
- Dependency injection through constructor parameters worked beautifully. 
- I was able to house my custom middle ware implementation in a separate class library with just 2 lines of initialization ( ServiceCollection.AddTransient and builder.UseMiddleWare)
- In the end there was no code clutter, very clear separation of concerns and DI based injection

I am glad to see that Microsoft corrected its course and did the right thing. Thank you.
